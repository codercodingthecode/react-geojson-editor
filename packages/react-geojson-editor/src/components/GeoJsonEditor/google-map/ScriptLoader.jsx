/* global google */
/* eslint-disable react/no-did-mount-set-state */
import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { fetchJsScript } from '../../../utils';

// withScriptjs
const LOADING_STATE_NONE = 'NONE';
const LOADING_STATE_BEGIN = 'BEGIN';
const LOADING_STATE_LOADED = 'LOADED';

// context is the global window.google object set when the script has been loaded
export const ScriptContext = React.createContext(null);

export class ScriptLoader extends Component {
  constructor(props) {
    super(props);
    this.handleLoaded = this.handleLoaded.bind(this);
  }
    state = {
      loadingState: LOADING_STATE_NONE,
      google: null,
    };

    componentDidMount() {
      const { scriptUrl } = this.props;
      const { loadingState } = this.state;
      if (loadingState !== LOADING_STATE_NONE) return;
      this.setState({ loadingState: LOADING_STATE_BEGIN }, () => {
        // @TODO have fetchJsScript provide google object in the callback
        // fetchJsScript(scriptUrl, google => this.handleLoaded(google))
        fetchJsScript(scriptUrl, () => this.handleLoaded(google));
      });
    }
    componentWillUnmount() {
      this.isUnmounted = true;
    }
    isUnmounted = false

    handleLoaded(google) {
      if (this.isUnmounted) {
        return;
      }
      this.setState({
        google,
        loadingState: LOADING_STATE_LOADED,
      });
    }

    render() {
      const { google, loadingState } = this.state;
      if (loadingState !== LOADING_STATE_LOADED) { return <div> Loading ... </div>; }

      if (!google) { return <div> Error Loading Script </div>; }

      return (
        <ScriptContext.Provider value={google}>
          {this.props.children}
        </ScriptContext.Provider>
      );
    }
}

ScriptLoader.propTypes = {
  children: PropTypes.node.isRequired,
  scriptUrl: PropTypes.string.isRequired,
};

export default ScriptLoader;
