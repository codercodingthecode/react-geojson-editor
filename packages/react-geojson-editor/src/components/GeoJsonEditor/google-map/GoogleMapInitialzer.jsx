/* eslint-disable react/no-unused-state */
/* eslint-disable react/no-did-mount-set-state */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { ScriptContext } from './ScriptLoader';

// withGoogleMap
export const GoogleMapContext = React.createContext({
  google: null,
  map: null,
});
export class GoogleMapInitialzer extends Component {
    static contextType = ScriptContext;

    constructor(props, context) {
      super(props, context);
      if (!context) { throw new Error('No google object found, please make sure you have loaded google map script'); }
      this.mapElement = React.createRef();
    }
    state = { google: null, map: null };

    componentDidMount() {
      const { center, zoom } = this.props;
      const google = this.context;
      const map = new google.maps.Map(this.mapElement.current, {
        center,
        zoom,
        fullscreenControl: false,
        streetViewControl: false,
        mapTypeControlOptions: {
          style: google.maps.MapTypeControlStyle.DROPDOWN_MENU,
        },
      });
      this.setState({ google, map });
    }

    componentDidUpdate(prevProps) {
      const { center } = this.props;
      if (prevProps.center !== center) { this.state.map.setCenter(center); }
    }

    render() {
      const { width, height } = this.props;
      const { map } = this.state;
      return (
        <div style={{ width, height }} ref={this.mapElement}>
          {
            map ?
              <GoogleMapContext.Provider value={this.state}>
                {this.props.children}
              </GoogleMapContext.Provider> :
              null
          }
        </div>
      );
    }
}

GoogleMapInitialzer.propTypes = {
  center: PropTypes.shape({
    lat: PropTypes.number,
    lng: PropTypes.number,
  }).isRequired,
  children: PropTypes.node.isRequired,
  // @note height needs to be greater than 0px
  // @source: https://developers.google.com/maps/documentation/javascript/adding-a-google-map
  height: PropTypes.number.isRequired,
  width: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]).isRequired,
  zoom: PropTypes.number.isRequired,
};

export default GoogleMapInitialzer;
