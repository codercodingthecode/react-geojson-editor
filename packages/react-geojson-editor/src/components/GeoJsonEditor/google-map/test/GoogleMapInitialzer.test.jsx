import React from 'react';

import { shallow, mount } from 'enzyme';
import { GoogleMapInitialzer } from '../GoogleMapInitialzer';
import { ScriptContext } from '../ScriptLoader';

describe('Button', () => {
  it('has not changed', () => {
    const props = {
      center: { lat: 0, lng: 0 },
      zoom: 1,
      height: 100,
      width: 100,
    };
    const wrapper = shallow(<GoogleMapInitialzer {...props}><div /></GoogleMapInitialzer>);
    expect(wrapper).toMatchSnapshot();
  });
  it('renders by creating a map', () => {
    const props = {
      center: { lat: 0, lng: 0 },
      zoom: 1,
      height: 100,
      width: 100,
    };
    const wrapper = mount(
      <ScriptContext.Provider value={google}>
        <GoogleMapInitialzer {...props}><div /></GoogleMapInitialzer>
      </ScriptContext.Provider>
    );

    expect(wrapper.state().map).toBeDefined();
  });
});
