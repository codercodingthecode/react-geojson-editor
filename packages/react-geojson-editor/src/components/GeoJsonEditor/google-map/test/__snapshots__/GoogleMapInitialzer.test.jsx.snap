// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button has not changed 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <GoogleMapInitialzer
    center={
      Object {
        "lat": 0,
        "lng": 0,
      }
    }
    height={100}
    width={100}
    zoom={1}
  >
    <div />
  </GoogleMapInitialzer>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": null,
      "style": Object {
        "height": 100,
        "width": 100,
      },
    },
    "ref": Object {
      "current": null,
    },
    "rendered": null,
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": null,
        "style": Object {
          "height": 100,
          "width": 100,
        },
      },
      "ref": Object {
        "current": null,
      },
      "rendered": null,
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
